### ISSUE TITLE
Enable rlib-only libstd build

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#56443]
For cross-building libstd (for use e.g. with [miri](https://github.com/solson/miri/)), it would be great if one could make an rlib-only libstd build. This is because building a dylib invokes the native linker, and that fails when the target architectures is too foreign and/or not supported by the installed C toolchain. (Currently, we build libstd both as rlib and dylib.)

So my goal was to equip libstd with a feature flag to control whether it is built as a dylib (next to the rlib) or not.  To this end, I patched [`collect_crate_types`](https://github.com/rust-lang/rust/blob/8660eba2b9bec5b0fe971b7281f79e79c2df2fae/src/librustc_driver/driver.rs#L1500) such that the crate types passed on the command-line and the one given as attributes are merged (instead of command-line taking precedence). That means I can use `#![cfg_attr(dylib, crate-type = "dylib")]` to make a `dylib` feature enable a dylib build. That seems to work, however, there is a problem: During bootstrap, rustbuild uses cargo's stamp file to determine which files to copy into the sysroot, and that stamp file now no longer includes the `libstd.so`, breaking the build.

I wonder if there is maybe a better way to achieve an rlib-only libstd build?

Cc @eddyb @alexcrichton any advice?

**Comment by eddyb** [comment]
cc @rust-lang/compiler This interaction between CLI flags and attributes seems really unfortunate.
I wonder if `Cargo` could benefit from a `--print` option for this, that would be handled after macro expansion, but it would have to happen at the same time as *compiling* the crate (as to avoid doing parsing and expansion twice), so it'd be more like `--emit=dep-info`.

**Comment by alexcrichton** [comment]
This is currently a limitation of the compiler and we can't conditionally produce a dynamic library. The CLI flags passed to `rustc` trump whatever is listed in the crate (attribute-wise), and there's no way to have Cargo conditionally produce a dynamic library for the standard library, unfortunately.

We could consider adding a feature to rustc for this though! (it's certainly been asked for before)

**Comment by RalfJung** [comment]
@alexcrichton I have a [branch](https://github.com/RalfJung/rust/commits/libstd-no-dylib) that makes CLI flags not trump attributes any more.

The problem is, as described in my first post: With that patch, rustubild no longer copies the `libstd.so` into the sysroot because it is not listed in the stamp file.

**Comment by alexcrichton** [comment]
Yeah crate type is tricky where it can't really be configured without Cargo's knowledge as Cargo moves around files and need to know what's being output. I think this would need to be a Cargo feature rather than a rustc feature to implement this

**Comment by RalfJung** [comment]
@alexcrichton suggested it might be possible to do at least a metadata-only build, using essentially `cargo check`.

I tried this, but at least for cross-building purposes that currently does not work because since compiler_builtins got moved out of the rustc workspace, the `c` feature is forced to be enabled. (It can be controlled in libstd, but not in liballoc which sets `rustc-dep-of-std`.)

**Comment by Ericson2314** [comment]
Can we just have the workspace root decide how things are built?

**Comment by RalfJung** [comment]
Looks like https://github.com/rust-lang/cargo/pull/7353 might be related or even solve this issue, but I have not taken a closer look yet.
